test/command/suite/sharding/logical_range_filter/no_index/condition/no_range.expected

```diff
@@ -74,6 +76,12 @@ logical_range_filter Logs timestamp --filter 'message == "Shutdown"'
     ]
   ]
 ]
+#|d| [logical_range_filter][range-index] <Logs_20150203>: the max number of unmatched records (0) <= the number of estimated records (1)
+#|d| [logical_range_filter][select] <Logs_20150203>: no range index
```

Comment:
When we cannot find any range index, logs for `use range index` and `do not use range index` are written when we cannot find any range index.

1. We log `#|d| [logical_range_filter][range-index] <Logs_20150203>: the max number of unmatched records (0) <= the number of estimated records (1)` when we decide to use range index in `use_range_index?`
2. After that, we log `#|d| [logical_range_filter][select] <Logs_20150203>: no range index` because we cannot find any range index.

I think this logging behavior is not smart but acceptable because it is only for debugging.
How do you feel that?
`#|d| [logical_range_filter][range-index] <Logs_20150203>: the max number of unmatched records (0) <= the number of estimated records (1)` is written by `use_range_index?` is determined to use `range_index`

Reply:
Umm... It will make debug difficult. `[range-index]` and `[select]` is exclusive for now. It means that we can know that we use a range index by `[range-index]` log. But we can't know that we use a range index by `[range-index]` log with this change. We also need confirm no `[select]` log to know whether a range index is used or not.

How about the following?

```diff
diff --git a/plugins/sharding/logical_range_filter.rb b/plugins/sharding/logical_range_filter.rb
index f46bd26f4..16b5f9bc4 100644
--- a/plugins/sharding/logical_range_filter.rb
+++ b/plugins/sharding/logical_range_filter.rb
@@ -252,52 +252,57 @@ module Groonga
           end
         end
 
-        def use_range_index?
+        def find_range_index
           use_range_index_parameter_message =
             "force by use_range_index parameter"
           case @context.use_range_index
           when true
-            return decide_use_range_index(true,
-                                          use_range_index_parameter_message,
-                                          __LINE__, __method__)
+            return find_range_index_raw(use_range_index_parameter_message,
+                                        __LINE__, __method__)
+            end
           when false
-            return decide_use_range_index(false,
-                                          use_range_index_parameter_message,
-                                          __LINE__, __method__)
+            log_use_range_index(false,
+                                use_range_index_parameter_message,
+                                __LINE__, __method__)
+            return nil
           end
 
           range_index_logical_parameter_message =
             "force by range_index logical parameter"
           case Parameters.range_index
           when :always
-            return decide_use_range_index(true,
-                                          range_index_logical_parameter_message,
-                                          __LINE__, __method__)
+            return find_range_index_raw(range_index_logical_parameter_message,
+                                        __LINE__, __method__)
           when :never
-            return decide_use_range_index(false,
-                                          range_index_logical_parameter_message,
-                                          __LINE__, __method__)
+            log_use_range_index(false,
+                                range_index_logical_parameter_message,
+                                __LINE__, __method__)
+            return nil
           end
 
           unless @context.dynamic_columns.empty?
             reason = "dynamic columns are used"
-            return decide_use_range_index(false, reason, __LINE__, __method__)
+            log_use_range_index(false, reason, __LINE__, __method__)
+            return nil
           end
 
           unless @context.post_filter.nil?
             reason = "post_filter is used"
-            return decide_use_range_index(false, reason, __LINE__, __method__)
+            log_use_range_index(false, reason, __LINE__, __method__)
+            return nil
           end
 
           unless @context.sort_keys.empty?
             reason = "sort_keys is used"
-            return decide_use_range_index(false, reason, __LINE__, __method__)
+            log_use_range_index(false, reason, __LINE__, __method__)
+            return nil
           end
 
           current_limit = @context.current_limit
           if current_limit < 0
             reason = "limit is negative: <#{current_limit}>"
-            return decide_use_range_index(false, reason, __LINE__, __method__)
+            log_use_range_index(false, reason, __LINE__, __method__)
+            return nil
           end
 
           required_n_records = @context.current_offset + current_limit
@@ -307,20 +312,19 @@ module Groonga
             reason = "the number of required records (#{required_n_records}) "
             reason << ">= "
             reason << "the number of records in shard (#{max_n_records})"
-            return decide_use_range_index(false, reason,
-                                          __LINE__, __method__)
+            log_use_range_index(false, reason,__LINE__, __method__)
+            return nil
           end
 
           threshold = @context.threshold
           if threshold <= 0.0
             reason = "threshold is negative: <#{threshold}>"
-            return decide_use_range_index(true, reason,
-                                          __LINE__, __method__)
+            return find_range_index_raw(reason, __LINE__, __method__)
           end
           if threshold >= 1.0
             reason = "threshold (#{threshold}) >= 1.0"
-            return decide_use_range_index(false, reason,
-                                          __LINE__, __method__)
+            log_use_range_index(false, reason, __LINE__, __method__)
+            return nil
           end
 
           table = @shard.table
@@ -332,7 +336,7 @@ module Groonga
                 @expression_builder.build_all(expression)
                 unless range_index_available_expression?(expression,
                                                          __LINE__, __method__)
-                  return false
+                  return nil
                 end
                 estimated_n_records = expression.estimate_size(table)
               end
@@ -344,7 +348,7 @@ module Groonga
               @expression_builder.build_partial_min(expression)
               unless range_index_available_expression?(expression,
                                                        __LINE__, __method__)
-                return false
+                return nil
               end
               estimated_n_records = expression.estimate_size(table)
             end
@@ -353,7 +357,7 @@ module Groonga
               @expression_builder.build_partial_max(expression)
               unless range_index_available_expression?(expression,
                                                        __LINE__, __method__)
-                return false
+                return nil
               end
               estimated_n_records = expression.estimate_size(table)
             end
@@ -362,7 +366,7 @@ module Groonga
               @expression_builder.build_partial_min_and_max(expression)
               unless range_index_available_expression?(expression,
                                                        __LINE__, __method__)
-                return false
+                return nil
               end
               estimated_n_records = expression.estimate_size(table)
             end
@@ -372,8 +376,8 @@ module Groonga
             reason = "the number of required records (#{required_n_records}) "
             reason << ">= "
             reason << "the number of estimated records (#{estimated_n_records})"
-            return decide_use_range_index(false, reason,
-                                          __LINE__, __method__)
+            log_use_range_index(false, reason, __LINE__, __method__)
+            return nil
           end
 
           max_n_unmatched_records =
@@ -383,8 +387,7 @@ module Groonga
             reason = "the max number of unmatched records (#{max_n_unmatched_records}) "
             reason << "<= "
             reason << "the number of estimated records (#{estimated_n_records})"
-            return decide_use_range_index(true, reason,
-                                          __LINE__, __method__)
+            return find_range_index_raw(reason, __LINE__, __method__)
           end
 
           hit_ratio = estimated_n_records / max_n_records.to_f
@@ -397,8 +400,11 @@ module Groonga
           reason = "hit ratio "
           reason << "(#{hit_ratio}=#{estimated_n_records}/#{max_n_records}) "
           reason << "#{relation} threshold (#{threshold})"
-          decide_use_range_index(use_range_index_by_hit_ratio, reason,
-                                 __LINE__, __method__)
+          if use_range_index_by_hit_ratio
+            find_range_index_raw(reason, __LINE__, __method__)
+          else
+            decide_use_range_index(false, reason, __LINE__, __method__)
+            nil
         end
 
         def range_index_available_expression?(expression, line, method_name)
diff --git a/plugins/sharding/stream_shard_executor.rb b/plugins/sharding/stream_shard_executor.rb
index 387d1f498..b66ecb250 100644
--- a/plugins/sharding/stream_shard_executor.rb
+++ b/plugins/sharding/stream_shard_executor.rb
@@ -3,7 +3,7 @@ module Groonga
     class StreamShardExecutor
       # Derived class must imprement methods below.
       # - execute_filter(range_index)
-      # - use_range_index?
+      # - find_range_index
 
       attr_reader :shard
       attr_writer :previous_executor
@@ -99,12 +99,7 @@ module Groonga
         @expression_builder = RangeExpressionBuilder.new(@shard_key,
                                                          @target_range)
         @expression_builder.filter = @filter
-        index_info = @shard_key.find_index(Operator::LESS)
-        if index_info
-          index = index_info.index
-          @context.referred_objects << index
-          @range_index = index if use_range_index?
-        end
+        @range_index = find_range_index
 
         if @context.dynamic_columns.have_initial?
           if @cover_type == :all
@@ -125,6 +120,19 @@ module Groonga
         @window = detect_window
       end
 
+      def find_range_index_raw(use_reason, line, method)
+        index_info = @shard_key.find_index(Operator::LESS)
+        if index_info.nil?
+          log_use_range_index(false, "no range index", __LINE__, __method__)
+          return nil
+        end
+
+        index = index_info.index
+        @context.referred_objects << index
+        log_use_range_index(true, use_reason, line, method)
+        index
+      end
+
       def ensure_filtered
         return if @filtered
 
@@ -171,7 +179,7 @@ module Groonga
         end
       end
 
-      def decide_use_range_index(use, reason, line, method)
+      def log_use_range_index(use, reason, line, method)
         message = "[#{@command_name}]"
         if use
           message << "[range-index] "
@@ -185,7 +193,6 @@ module Groonga
                    line,
                    method.to_s,
                    message)
-        use
       end
 
       def apply_post_filter(table)
```

Reply:
Thank you. It looks good.
I have modified the code with this approach.